src/nvim/os/shell.c

```diff
@@ -295,7 +309,8 @@ static void system_data_cb(Stream *stream, RBuffer *buf, void *data, bool eof)
   dbuf->len += nread;
 }
 
-static void out_data_cb(Stream *stream, RBuffer *buf, void *data, bool eof)
+static void out_data_cb(Stream *stream, RBuffer *buf, size_t count, void *data,
+    bool eof)
 {
   RBUFFER_UNTIL_EMPTY(buf, ptr, len) {
```

Comment:
If we use a shell command with a lot of output, like `:!ls / -R`, output will hang.
Replacing this macro with 
```c
  if (count == 0) {
    return;
  }
  size_t cnt;
  char *ptr = rbuffer_read_ptr(buf, &cnt);
  size_t written = write_output(ptr, count, false, 1);
  rbuffer_consumed(buf, count);

```
works for this case.